cli/flox-rust-sdk/src/providers/build.rs

```diff
@@ -161,3 +161,435 @@ fn read_output_to_channel(
+        let env_path = env.parent_path().unwrap();
+
+        // This file is accessible from an impure build.
+        fs::write(env_path.join(&file_name), &file_content).unwrap();
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+    }
+
+    #[test]
+    #[ignore] // TODO: Intermittently uses cache.
+    fn build_cache_sandbox_pure_uses_cache() {
+        let package_name = String::from("foo");
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            sandbox = "pure"
+            command = """
+                mkdir $out
+                echo -n {file_content} > $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_first = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_second = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_eq!(dir_second, dir_first);
+    }
+
+    #[test]
+    #[ignore] // TODO: Intermittently uses cache.
+    fn build_cache_sandbox_pure_cache_can_be_invalidated() {
+        let package_name = String::from("foo");
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            sandbox = "pure"
+            command = """
+                mkdir $out
+                echo -n {file_content} > $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_first = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        let cache_dir = cache_dir(&env_path, &package_name);
+        assert!(cache_dir.exists());
+        fs::remove_file(cache_dir).unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_second = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_ne!(dir_second, dir_first);
+    }
```

Comment:
This looks like a misunderstanding of the caching feature.
For me this test passes but only because of the cache store path issue of the test above.
Notably, the test _still_ passes when we remove the `fs::remove_file(cache_dir).unwrap();`.

Optimally even if we removed the cache file, the resulting store path shouldn't change at least when applying nix semantics.
The caching applied by `flox-build.mk` might be a higher level mechanism after all.

Regardless this test should probably rather ensure that an action that produces the cached value is not executed twice.

Reply:
Follow up in person? Not sure if there's a question in here ...

Reply:
The main question is, how do we tell that a cache is working or in this case can be invalidated.

We currently do so by asserting that the resulting store link is the same (which tracks with the understanding of caching of native nix).
Apparently that is not the case, and it takes 3 builds for that behavior to show actually.

I sketched a different example in the other comments, to test caching that should apply here as well.


Somewhat orthogonally, maybe you can comment on how the sandbox cache works since that is purely implemented in the makefile and an extension of plain store path caching provided by nix.